File: roms/super_real_mahjong_part2.py
#
#	Super Real Mahjong Part 2
#

from base64 import b64encode
from io import BytesIO
from math import ceil
from sys import argv
from zipfile import ZipFile
from PIL import Image

def merge(even, odd):
    ret = bytearray(len(even) * 2)
    ret[0::2] = even
    ret[1::2] = odd
    return bytes(ret)

with ZipFile(argv[1]) as z:
    prg = merge(z.read('uco-2.17'), z.read('uco-3.18'))
    gfx = z.read('ubo-4.60') + z.read('ubo-5.61') + merge(z.read('uco-8.64'), z.read('uco-9.65')) + z.read('ubo-6.62') + z.read('ubo-7.63') + merge(z.read('uco-10.66'), z.read('uco-11.67'))
    voi = z.read('uco-1.19')
    color_h = z.read('uc-1o.12')
    color_l = z.read('uc-2o.13')

rom = prg + gfx + voi + color_h + color_l

def pngstring(a):
    w = 1024
    img = Image.new('L', (w, ceil(len(a) / w)))
    img.putdata(a)
    buf = BytesIO()
    img.save(buf, 'PNG')
    return b64encode(buf.getvalue())

with open(argv[2], 'wb') as f:
    f.write(b'export const ROM = \'data:image/png;base64,' + pngstring(rom) + b'\';\n')

File: roms/super_real_mahjong_part2.js
/*
 *	Super Real Mahjong Part 2
 */

const fs = require('fs');
const {createCanvas, createImageData} = require('canvas');
const {Zlib} = require('zlibjs/bin/unzip.min');
Zlib.Unzip.prototype.read = Zlib.Unzip.prototype.decompress;

const cat = (...args) => {
	const array = new Uint8Array(args.reduce((a, b) => a + b.length, 0));
	return args.reduce((a, b) => (array.set(b, a), a + b.length), 0), array;
};
const merge = (...args) => {
	const array = new Uint8Array(args.reduce((a, b) => a + b.length, 0)), step = args.length;
	return args.forEach((a, i) => a.forEach((e, j) => array[j * step + i] = e)), array;
};

const z = new Zlib.Unzip(fs.readFileSync(process.argv[2]));
const prg = merge(z.read('uco-2.17'), z.read('uco-3.18'));
const gfx = cat(z.read('ubo-4.60'), z.read('ubo-5.61'), merge(z.read('uco-8.64'), z.read('uco-9.65')), z.read('ubo-6.62'), z.read('ubo-7.63'), merge(z.read('uco-10.66'), z.read('uco-11.67')));
const voi = z.read('uco-1.19');
const color_h = z.read('uc-1o.12');
const color_l = z.read('uc-2o.13');

const rom = cat(prg, gfx, voi, color_h, color_l);

const pngString = array => {
	const w = 1024, h = Math.ceil(array.byteLength / w), canvas = createCanvas(w, h);
	const palette = new Uint8ClampedArray(new Uint32Array(256).map((e, i) => i | 0xff000000).buffer);
	canvas.getContext('2d', {pixelFormat: 'A8'}).putImageData(createImageData(new Uint8ClampedArray(array.buffer), w, h), 0, 0);
	return Buffer.from(canvas.toBuffer('image/png', {compressionLevel: 9, filters: canvas.PNG_FILTER_NONE, palette})).toString('base64');
};

fs.writeFileSync(process.argv[3], `export const ROM = 'data:image/png;base64,${pngString(rom)}';\n`);
